### ISSUE TITLE
CC, CXX: .pythranrc + env variables, who win?

### ISSUE THREAD

**Issue by paugier** [serge-sans-paille/pythran#1047]
It seems to me that when both CC/CXX environment variables and .pythranrc are set, the content of the environment variables is used.

In the doc, it is not clear who win in this case:

> Wants to try your own compiler? Update the CC and CXX fields from your pythranrc, or set the same env variables to the right compilers.

On Intel environment, it seems to me that it would be good that 

```
.pythranrc > env variables
```

because CC and CXX can be used for other things (to use icc).

If we want to keep environment variables stronger than the values in .pythranrc, we can have 

```
PYTHRAN_CC / PYTHRAN_CPP > .pythranrc > CC/CPP
```

**Comment by serge-sans-paille** [comment]
mmmh, while I understand your issue, the generally accepted rule is to have env variable take precedence over config file, as in:

- https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/environment.html
- https://docs.docker.com/compose/env-file/

so we definitievely want

```
env variables > .pythranrc
```

In addition to common practice, the rational seems to be: you can change an env variable per call, while you don't change settings that often.

It would be an API change to rename CC/CXX to PYTHRAN_*, and I'm not sure I want to go that way.
Still there is hope for you :-)

If you want to set the compiler used by pythran while CC/CXX are already set, say by ``iccsetvars.sh`` then add an alias:

```
alias pythran="CC=gcc CXX=gcc pythran"
```

and if you want ``PYTHRAN_CC`` to override ``CC``, a simple wrapper like

```
!/bin/sh
PYTHRAN="/path/to/pythran"
if [ -n "$PYTHRAN_CC" ]; then
    export CC=$PYTHRAN_CC 
fi
$PYTHRAN $*
```

should do the trick

**Comment by serge-sans-paille** [comment]
#1050 clarifies the situation.

**Comment by paugier** [comment]
Your proposed solutions would work when using the command line pythran but not when Pythran is used in a setup.py.  In this case, I don't see how one can build a package using Pythran if CC/CXX point towards the Intel compiler. That's why I proposed that Pythran would be sensible both to CC and PYTHRAN_CC with the following precedence:

```PYTHRAN_CC / PYTHRAN_CXX > .pythranrc > CC/CXX```

But anyway, now that it is clear in the doc that `CC/CXX > .pythranrc`, it's acceptable.

But people (and I) will need to write warnings in the documentation of packages using Pythran, since in clusters, intel modules tend to define CC/CXX so users can be surprized to get a compile error doing (with a correct .pythranrc):

```
module load intel/17.2
pip install a_package_using_pythran
```

**Comment by serge-sans-paille** [comment]
What about 

```
PYTHRANCC > CC > .pythranrc
```

would that solve yourt issue?

**Comment by paugier** [comment]
It would solve an issue but then if we introduce another env variable (PYTHRANCC), I don't understand the advantage of 
```PYTHRANCC > CC > .pythranrc``` 
compare to 
```PYTHRANCC > .pythranrc > CC``` 

I see one advantage of `PYTHRANCC > .pythranrc > CC`, which is that we avoid this strange behavior for people who correctly set .pythranrc and don't know about CC/CXX variables and still cannot install a package using Pythran. The users get a unclear compilation error and don't understand what happen. For an example, see https://bitbucket.org/fluiddyn/fluidfft/issues/16

If we have PYTHRANCC + CC, one can easily override the config file setting per call.

If we have PYTHRANCC + CC, we don't rename CC/CXX to PYTHRAN_* since CC/CXX are still taken into account.

**Comment by serge-sans-paille** [comment]
The issue youmention makes sense. Still, CC > config is the defacto standard... I don't see any easy solution to this, apart from having pythran-genertaed code be icc compilant :-)

Documentation merged in #1050

**Comment by paugier** [comment]
> apart from having pythran-genertaed code be icc compilant :-)

Sure it would be the best solution!

I close this issue since the doc is now explicit. Users just have to be careful. :-)
